src/api/data/tree/importPlan.ts

```diff
@@ -191,50 +191,38 @@ const replaceRefWithId =
 
 export const validatePlanContents =
   (logger: Logger) =>
-  async (planPath: string, planContents: unknown): Promise<DataPlanPartFilesOnly[]> => {
+  (planPath: string, planContents: unknown): Promise<DataPlanPartFilesOnly[]> => {
     const childLogger = logger.child('validatePlanContents');
-    const planSchema = path.join(
-      path.dirname(fileURLToPath(import.meta.url)),
-      '..',
-      '..',
-      '..',
-      '..',
-      'schema',
-      'dataImportPlanSchema.json'
-    );
+    const parseResult = DataImportPlanArraySchema.safeParse(planContents);
+    if (!parseResult.success) {
+      const errMsg: string = parseResult.error.issues
+        .map((err) => `${err.path.join('.')}:\n  ${err.message}`)
+        .join('\n');
+      throw messages.createError('error.InvalidDataImport', [planPath, errMsg]);
+    }
 
-    const val = new SchemaValidator(childLogger, planSchema);
-    try {
-      await val.validate(planContents as AnyJson);
-      const output = planContents as DataPlanPartFilesOnly[];
-      if (hasRefs(output)) {
-        childLogger.warn(
-          "The plan contains the 'saveRefs' and/or 'resolveRefs' properties.  These properties will be ignored and can be removed."
-        );
-      }
-      if (!hasOnlySimpleFiles(output)) {
-        throw messages.createError('error.NonStringFiles');
-      }
-      return planContents as DataPlanPartFilesOnly[];
-    } catch (err) {
-      if (err instanceof Error && err.name === 'ValidationSchemaFieldError') {
-        throw messages.createError('error.InvalidDataImport', [planPath, err.message]);
-      } else if (err instanceof Error) {
-        throw SfError.wrap(err);
-      }
-      throw err;
+    const output: DataImportPlanSchemaType[] = parseResult.data;
+
+    if (hasRefs(output)) {
+      childLogger.warn(
+        "The plan contains the 'saveRefs' and/or 'resolveRefs' properties.  These properties will be ignored and can be removed."
+      );
+    }
+    if (!hasOnlySimpleFiles(output)) {
+      throw messages.createError('error.NonStringFiles');
     }
+    return Promise.resolve(planContents as DataPlanPartFilesOnly[]);
```

Comment:
I will almost never approve an `as`sertion in /src (I understand there's been one here for a really long time)

If we don't want to allow the DataPlanPart side of the TS union then let's block it, or manipulate it into the string[] that we expect.

The compiler here is trying to tell you that the types don't match the result type (zod definitely makes that clearer) and `as` is sweeping that under the rug.

Reply:
Or let me back up...the output of zod.parse/safeParse is a validated, **typed** object (why the TS world loves it so much).  And you can use `.infer` to get that from the schema like https://zod.dev/basics?id=inferring-types

I'd suggest looking through the consuming code to see whether it handles both scenarios (string[] vs. DataImportFile) and
- if yes, account for both types in this fn's return type (simplified if you return the `.infer`)
- if not, then we massage it here to return the expected type.

Reply:
we should also be able to get rid of this assertion above since it's being passed `unknown` for proper validation now.  

`        (await JSON.parse(fs.readFileSync(resolvedPlanPath, 'utf8'))) as DataPlanPartFilesOnly[]
`